src/lib/matchers.rs

```diff
@@ -216,19 +281,264 @@ impl<T: Clone> MappingStore<T> {
+                if z != v {
+                    let k = self.find_pos(x, &new_mappings, &to_in_order);
+                    let mut mov = Move::new(w, z, k);
+                    debug!("Edit script: MOV {} {} Parent: {} {}",
+                           self.from_arena[w].label,
+                           self.from_arena[w].value,
+                           self.to_arena[z].label,
+                           self.to_arena[z].value);
+                    mov.apply(&mut self.from_arena)?;
+                    script.push(mov);
+                }
+            }
+            // GumTree code has these lines also?
+            // from_in_order.insert(w);
+            // to_in_order.insert(x);
+            self.align_children(w,
+                                x,
+                                &mut script,
+                                &new_mappings,
+                                &mut from_in_order,
+                                &mut to_in_order)?;
+        }
+        // Delete phase.
+        let mut actions = EditScript::new();
+        for w in root.post_order_traversal(&self.from_arena) {
+            if !new_mappings.contains_from(&w) {
+                debug!("Edit script: DEL {} {}",
+                       self.from_arena[w].label,
+                       self.from_arena[w].value);
+                let del = Delete::new(w);
+                script.push(del);
+                actions.push(del);
+            }
+        }
+        actions.apply(&mut self.from_arena)?;
+        // Add new mappings to the existing store.
+        debug!("Append new mappings to store.");
+        for (from, to) in new_mappings.from {
+            if !self.contains_from(&from) {
+                debug!("New mapping due to edit script: {} -> {}", from, to);
+                self.push(from, to, MappingType::EDIT);
+            }
+        }
+        // debug_assert!(self.is_isomorphic(root, root));
+        Ok(script)
+    }
+
+    fn align_children(&mut self,
+                      w: NodeId,
+                      x: NodeId,
+                      script: &mut EditScript<T>,
+                      new_mappings: &TemporaryMappingStore,
+                      from_in_order: &mut HashSet<NodeId>,
+                      to_in_order: &mut HashSet<NodeId>)
+                      -> Result<(), ArenaError> {
+        debug!("align_children({}, {})", w, x);
+        for child in w.children(&self.from_arena) {
+            from_in_order.remove(&child);
+        }
+        for child in x.children(&self.to_arena) {
+            to_in_order.remove(&child);
+        }
+        let mut s1: Vec<NodeId> = vec![];
+        for child in w.children(&self.from_arena) {
+            if new_mappings.contains_from(&child) {
+                let mapped = new_mappings.get_to(&child).unwrap();
+                if x.children(&self.to_arena)
+                       .collect::<Vec<NodeId>>()
+                       .contains(&mapped) {
```

Comment:
It's probably quicker to use `find` (or whatever it's called) rather than `collect` and `contains` (since `collect` is going to malloc memory etc.).